File: Python/XGB-PClim_predict.py
#!/usr/bin/env python3
"""
XGB-PClim Climate Prediction -- Python Implementation
=====================================================
Predict mean annual temperature (MAT) and mean annual precipitation (MAP)
from major-oxide geochemistry of paleosols using pre-trained XGBoost models.

Usage
-----
    python XGB-PClim_predict.py
    python XGB-PClim_predict.py --data path/to/input.csv --output path/to/output.csv
"""

import argparse
import sys
from pathlib import Path

import joblib
import numpy as np
import pandas as pd
from scipy.special import inv_boxcox


# ==========================================================
# Required for unpickling the temperature pipeline
# ==========================================================
def identity(x):
    """Identity function used as a no-op transformer."""
    return x


# ==========================================================
# Required oxide columns
# ==========================================================
OXIDE_COLUMNS = [
    "CaO", "MgO", "Na2O", "K2O", "Al2O3",
    "Fe2O3", "SiO2", "MnO", "TiO2", "P2O5",
]


# ==========================================================
# Default paths
# ==========================================================
DEFAULT_DATA_PATH = Path(__file__).parent / "data" / "input.csv"

DEFAULT_PRECIP_MODEL = (
    Path(__file__).parent
    / "models"
    / "precipitation_model_artifacts.joblib"
)

DEFAULT_TEMP_MODEL = (
    Path(__file__).parent
    / "models"
    / "temperature_model_artifacts.joblib"
)

DEFAULT_OUTPUT = (
    Path(__file__).parent.parent
    / "results"
    / "output.csv"
)


# ==========================================================
# Model uncertainties
# ==========================================================
MAT_UNCERTAINTY = 4.1      # °C
MAP_UNCERTAINTY = 322    # mm/year


# ==========================================================
# Load models
# ==========================================================
def load_models(precip_path: Path, temp_path: Path):
    """Load precipitation and temperature model artifacts."""

    precip_art = joblib.load(precip_path)
    temp_art = joblib.load(temp_path)

    return precip_art, temp_art


# ==========================================================
# Prediction function
# ==========================================================
def predict(
    data: pd.DataFrame,
    precip_art: dict,
    temp_art: dict
) -> pd.DataFrame:
    """
    Run MAT and MAP predictions on input data.

    Parameters
    ----------
    data : pd.DataFrame
        Input dataframe containing oxide columns.

    precip_art : dict
        Precipitation model artifacts.

    temp_art : dict
        Temperature model artifacts.

    Returns
    -------
    pd.DataFrame
        Input dataframe with prediction columns appended.
    """

    oxide_cols = precip_art["oxide_columns"]

    lam = precip_art["fitted_lambda"]
    shift = precip_art["shift"]

    # ------------------------------------------------------
    # Keep rows with complete oxide data
    # ------------------------------------------------------
    mask = data[oxide_cols].notnull().all(axis=1)

    X = data.loc[mask, oxide_cols]

    # ======================================================
    # MAP prediction
    # ======================================================
    precip_bc = precip_art["pipeline"].predict(X)

    precip = inv_boxcox(precip_bc, lam) - shift

    precip = np.maximum(precip, 0.0)

    # ======================================================
    # MAT prediction
    # ======================================================
    temp = temp_art["pipeline"].predict(X)

    # ======================================================
    # Round predictions
    # ======================================================
    temp = np.round(temp, 1)

    precip = np.round(precip, 0)
    precip = precip.astype(int)
    # ======================================================
    # Calculate uncertainty ranges
    # ======================================================
    temp_min = np.round(temp - MAT_UNCERTAINTY, 1)

    temp_max = np.round(temp + MAT_UNCERTAINTY, 1)

    precip_min = np.round(
        np.maximum(precip - MAP_UNCERTAINTY, 0),
        0
    )

    precip_max = np.round(
        precip + MAP_UNCERTAINTY,
        0
    )

    # ======================================================
    # Create result dataframe
    # ======================================================
    result = data.copy()

    # MAT columns
    result["MAT_Best"] = np.nan
    result["MAT_Min"] = np.nan
    result["MAT_Max"] = np.nan

    # MAP columns
    result["MAP_Best"] = np.nan
    result["MAP_Min"] = np.nan
    result["MAP_Max"] = np.nan

    # ======================================================
    # Store predictions
    # ======================================================
    result.loc[mask, "MAT_Best"] = temp
    result.loc[mask, "MAT_Min"] = temp_min
    result.loc[mask, "MAT_Max"] = temp_max

    result.loc[mask, "MAP_Best"] = precip
    result.loc[mask, "MAP_Min"] = precip_min
    result.loc[mask, "MAP_Max"] = precip_max

    return result


# ==========================================================
# Main
# ==========================================================
def main():

    parser = argparse.ArgumentParser(
        description=(
            "XGB-PClim: Predict paleoclimate from "
            "paleosol geochemistry"
        ),
    )

    parser.add_argument(
        "--data",
        type=Path,
        default=DEFAULT_DATA_PATH,
        help="Path to input CSV with oxide columns",
    )

    parser.add_argument(
        "--precip-model",
        type=Path,
        default=DEFAULT_PRECIP_MODEL,
        help="Path to precipitation model .joblib file",
    )

    parser.add_argument(
        "--temp-model",
        type=Path,
        default=DEFAULT_TEMP_MODEL,
        help="Path to temperature model .joblib file",
    )

    parser.add_argument(
        "--output",
        type=Path,
        default=DEFAULT_OUTPUT,
        help="Path for output CSV",
    )

    args = parser.parse_args()

    # ======================================================
    # Validate input files
    # ======================================================
    for label, path in [
        ("Data", args.data),
        ("Precipitation model", args.precip_model),
        ("Temperature model", args.temp_model),
    ]:

        if not path.exists():

            print(
                f"Error: {label} file not found:\n{path}",
                file=sys.stderr
            )

            sys.exit(1)

    # ======================================================
    # Load input data
    # ======================================================
    data = pd.read_csv(args.data)

    # Remove accidental unnamed columns
    data = data.loc[
        :,
        ~data.columns.str.contains("^Unnamed")
    ]

    # ======================================================
    # Load models
    # ======================================================
    precip_art, temp_art = load_models(
        args.precip_model,
        args.temp_model
    )

    # ======================================================
    # Run predictions
    # ======================================================
    result = predict(
        data,
        precip_art,
        temp_art
    )

    # ======================================================
    # Save output
    # ======================================================
    args.output.parent.mkdir(
        parents=True,
        exist_ok=True
    )

    result.to_csv(
        args.output,
        index=False
    )

    # ======================================================
    # Console output
    # ======================================================
    print(f"\nPredictions saved to:\n{args.output}")

    print(
        f"Rows predicted: "
        f"{result['MAT_Best'].notna().sum()} / {len(result)}"
    )


# ==========================================================
# Run
# ==========================================================
if __name__ == "__main__":
    main()

File: R/XGB-PClim_predict.R
#!/usr/bin/env Rscript
# ==========================================================
# XGB-PClim Climate Prediction -- R Implementation
# ==========================================================
# Predict mean annual temperature (MAT) and mean annual
# precipitation (MAP) from major-oxide geochemistry of
# paleosols using pre-trained XGBoost models.
#
# Usage:
#   Rscript XGB-PClim_predict.R
#   Rscript XGB-PClim_predict.R --data path/to/input.csv --output path/to/output.csv
#
# Recommended project structure:
#
# XGB-PClim/
# ├── R/
# │   ├── XGB-PClim_predict.R
# │   ├── models/
# │   └── data/
# ├── results/
# ==========================================================

# --- Dependencies --------------------------------------------------------

required_packages <- c("xgboost")

for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg, repos = "https://cloud.r-project.org")
  }
}
library(xgboost)
# ---------------------------------------------------------
# Set working directory to script location
# ---------------------------------------------------------
args <- commandArgs(trailingOnly = FALSE)
script_path <- sub("--file=", "", args[grep("--file=", args)])
if (length(script_path) > 0) {
  setwd(dirname(normalizePath(script_path)))
}
cat("Working directory:\n")
print(getwd())

# --- Configuration -------------------------------------------------------

OXIDE_COLUMNS <- c(
  "CaO", "MgO", "Na2O", "K2O", "Al2O3",
  "Fe2O3", "SiO2", "MnO", "TiO2", "P2O5"
)

# --- Default paths -------------------------------------------------------

default_paths <- list(
  data          = "data/input.csv",
  temp_model    = "models/temperature_xgb_model.json",
  temp_mean     = "models/temp_scaler_mean.csv",
  temp_scale    = "models/temp_scaler_scale.csv",
  prec_model    = "models/precipitation_xgb_model.json",
  prec_mean     = "models/precp_scaler_mean.csv",
  prec_scale    = "models/precp_scaler_scale.csv",
  prec_boxcox   = "models/precp_boxcox_parameters.csv",
  output        = "../results/output_R.csv"
)

# --- Parse command-line arguments -----------------------------------------
parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  paths <- default_paths
  
  i <- 1
  while (i <= length(args)) {
    if (args[i] == "--data" && i < length(args)) {
      paths$data <- args[i + 1]; i <- i + 2
    } else if (args[i] == "--output" && i < length(args)) {
      paths$output <- args[i + 1]; i <- i + 2
    } else {
      i <- i + 1
    }
  }
  paths
}

# --- Helper: inverse Box-Cox transform -----------------------------------
inv_boxcox <- function(y, lambda) {
  if (abs(lambda) < 1e-10) {
    exp(y)
  } else {
    (lambda * y + 1)^(1 / lambda)
  }
}

# --- Main -----------------------------------------------------------------
main <- function() {
  paths <- parse_args()
  # Print detected project root
  cat("Working directory:", getwd(), "\n")
  # Validate inputs
  required <- c("data", "temp_model", "temp_mean", "temp_scale",
                "prec_model", "prec_mean", "prec_scale", "prec_boxcox")
  for (key in required) {
    if (!file.exists(paths[[key]])) {
      stop(sprintf("File not found: %s (%s)", paths[[key]], key))
    }
  }
  
  # --- Load data ----------------------------------------------------------
  data <- read.csv(paths$data, stringsAsFactors = FALSE)
  
  # Check oxide columns exist
  missing_cols <- setdiff(OXIDE_COLUMNS, names(data))
  if (length(missing_cols) > 0) {
    stop(paste("Missing oxide columns:", paste(missing_cols, collapse = ", ")))
  }
  
  # --- Load models and parameters -----------------------------------------
  temp_model <- xgb.load(paths$temp_model)
  prec_model <- xgb.load(paths$prec_model)
  
  temp_mean  <- read.csv(paths$temp_mean)$mean
  temp_scale <- read.csv(paths$temp_scale)$scale
  prec_mean  <- read.csv(paths$prec_mean)$mean
  prec_scale <- read.csv(paths$prec_scale)$scale
  
  prec_bc_params <- read.csv(paths$prec_boxcox)
  lambda <- prec_bc_params$boxcox_lambda
  shift  <- prec_bc_params$shift
  
  # --- Identify complete rows ---------------------------------------------
  mask <- complete.cases(data[, OXIDE_COLUMNS, drop = FALSE])
  
  # --- Temperature prediction ---------------------------------------------
  X_temp <- as.matrix(data[mask, OXIDE_COLUMNS])
  X_temp <- sweep(X_temp, 2, temp_mean, "-")
  X_temp <- sweep(X_temp, 2, temp_scale, "/")
  temp_pred <- predict(temp_model, X_temp)
  
  # --- Precipitation prediction -------------------------------------------
  X_prec <- as.matrix(data[mask, OXIDE_COLUMNS])
  X_prec <- sweep(log1p(X_prec), 2, prec_mean, "-")
  X_prec <- sweep(X_prec, 2, prec_scale, "/")
  prec_pred_bc <- predict(prec_model, X_prec)
  prec_pred <- inv_boxcox(prec_pred_bc, lambda) - shift
  prec_pred <- pmax(prec_pred, 0)
  
  # --- Assemble output ---------------------------------------------------- 
  temp_uncertainty <- 4.1   # example: ±1.8 °C
  prec_uncertainty <- 317    # example: ±95 mm/year
  temp_pred <- round(temp_pred, 1)
  prec_pred <- round(prec_pred, 0)
  
  data$MAT_Best <- NA_real_
  data$MAT_Min  <- NA_real_
  data$MAT_Max  <- NA_real_

  data$MAP_Best <- NA_real_
  data$MAP_Min  <- NA_real_
  data$MAP_Max  <- NA_real_

# --- Fill predictions -------------------------------------------------

  data$MAT_Best[mask] <- temp_pred
  data$MAT_Min[mask]  <- round(temp_pred - temp_uncertainty, 1)
  data$MAT_Max[mask]  <- round(temp_pred + temp_uncertainty, 1)

  data$MAP_Best[mask] <- prec_pred
  data$MAP_Min[mask]  <- round(pmax(prec_pred - prec_uncertainty, 0), 0)
  data$MAP_Max[mask]  <- round(prec_pred + prec_uncertainty, 0)
  
  # --- Save ---------------------------------------------------------------
  output_dir <- dirname(paths$output)
  if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)
  write.csv(data, paths$output, row.names = FALSE)
  
  
  # --- Summary ----------------------------------------------------------
  
  cat(sprintf("Predictions saved to: %s\n", paths$output))
  cat(sprintf("  Rows predicted: %d / %d\n", sum(mask), nrow(data)))
}

main()
